includes/sanitizers/class-amp-srcset-sanitizer.php

```diff
@@ -0,0 +1,123 @@
+<?php
+/**
+ * Class AMP_Srcset_Sanitizer.
+ *
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Srcset_Sanitizer
+ *
+ * Sanitizes the `srcset` attribute of elements.
+ *
+ * @internal
+ * @since 2.0.2
+ */
+class AMP_Srcset_Sanitizer extends AMP_Base_Sanitizer {
+
+	/**
+	 * Sanitize the HTML contained in the DOMDocument received by the constructor
+	 */
+	public function sanitize() {
+		$attribute_query = $this->dom->xpath->query( '//*/@srcset' );
+
+		if ( 0 === $attribute_query->length ) {
+			return;
+		}
+
+		foreach ( $attribute_query as $attribute ) {
+			/** @var DOMAttr $attribute */
+			if ( ! empty( $attribute->value ) ) {
+				$this->sanitize_srcset_attribute( $attribute );
+			}
+		}
+	}
+
+	/**
+	 * Parses the `srcset` attribute and validates each image candidate defined.
+	 *
+	 * Validation errors will be raised if the attribute value is malformed or contains image candidates
+	 * with the same width or pixel density defined.
+	 *
+	 * @param DOMAttr $attribute Srcset attribute.
+	 */
+	private function sanitize_srcset_attribute( DOMAttr $attribute ) {
+		$srcset = $attribute->value;
+
+		// Regex below is adapted from the JS validator. See https://github.com/ampproject/amphtml/blob/5fcb29a41d06867b25ed6aca69b4aeaf96456c8c/validator/js/engine/parse-srcset.js#L72-L81.
+		$matched = preg_match_all( '/\s*(?:,\s*)?(?<url>[^,\s]\S*[^,\s])\s*(?<dimension>[\d]+.?[\d]*[wx])?\s*(?:(?<comma>,)\s*)?/', $srcset, $matches );
```

Comment:
The regex should go into a class constant, with its "documentation" (the comment line here) as a docblock.

Otherwise, this is a string that is freshly built on every iteration, instead of reusing the constant. Later versions of PHP are smart enough to optimize automatically, but anything below 7.1 (iirc) isn't.

Reply:
There seems to be an error in the regex.
For the dimension, I think the `.` was meant to be escaped to match float multiples (i.e. `1.5x`).
As it is not escaped, it will not only match floats, but basically anything that ends in `w` or `x`. So it will match `500px` for example, because the `.?` matches against the `p`.
I think this would be more accurate:
```suggestion
		$matched = preg_match_all( '/\s*(?:,\s*)?(?<url>[^,\s]\S*[^,\s])\s*(?<dimension>[\d]+(?:\.[\d]+)?[wx])?\s*(?:(?<comma>,)\s*)?/', $srcset, $matches );
```

Reply:
Upstream PR to fix the regex: https://github.com/ampproject/amphtml/pull/30252

Reply:
> The regex should go into a class constant, with its "documentation" (the comment line here) as a docblock.
> 
> Otherwise, this is a string that is freshly built on every iteration, instead of reusing the constant. Later versions of PHP are smart enough to optimize automatically, but anything below 7.1 (iirc) isn't.

This got "outdated" now but still applies for performance reasons.

Reply:
Looking back at the [spec definition](https://html.spec.whatwg.org/multipage/images.html#width-descriptor) of a `srcset attribute`, it states that:

- a width descriptor is a number greater than zero representing the width descriptor value, and a U+0077 LATIN SMALL LETTER W character
- a pixel density descriptor is a valid floating-point number giving a number greater than zero representing the pixel density descriptor value, and a U+0078 LATIN SMALL LETTER X character

So float numbers should only match with the `x` character, and integers can either be matched with the `w` or `x` character.

Reply:
So the regex should more look like:

```php
$matched = preg_match_all( '/\s*(?:,\s*)?(?<url>[^,\s]\S*[^,\s])\s*(?<dimension>[\d]+[wx]|[\d]+\.[\d]+x)?\s*(?:(?<comma>,)\s*)?/', $srcset, $matches );
```

Reply:
Will you provide that feedback to https://github.com/ampproject/amphtml/pull/30252? Or is it not relevant upstream?

Reply:
Yes, it would be relevant upstream as well, will add my feedback.

Reply:
I think the regex is not quite right, @pierlon, as it makes the dimension optional. The question mark after the `(?<dimension>...)` should be stripped. I'll adapt the PR.

Reply:
Also, if you want to go with the exact definition you've posted, we'd need to exclude zero as a value.

Reply:
Upstream PR was merged.